pkg/controller/deactivation/deactivation_controller.go

```diff
@@ -149,10 +167,60 @@ func (r *ReconcileDeactivation) Reconcile(request reconcile.Request) (reconcile.
 	logger.Info("user account time values", "deactivation timeout duration", deactivationTimeout, "provisionedTimestamp", provisionedTimestamp)
 
 	timeSinceProvisioned := time.Since(provisionedTimestamp.Time)
-	if timeSinceProvisioned < deactivationTimeout {
+
+	deactivatingNotificationTimeout := time.Duration((deactivationTimeoutDays-config.Deactivation.DeactivatingNotificationDays)*24) * time.Hour
+
+	if timeSinceProvisioned < deactivatingNotificationTimeout {
+		// It is not yet time to send the deactivating notification so requeue until it will be time to send it
+		requeueAfterTimeToNotify := deactivatingNotificationTimeout - timeSinceProvisioned
+		logger.Info("requeueing request", "RequeueAfter", requeueAfterTimeToNotify,
+			"Expected deactivating notification date/time", time.Now().Add(requeueAfterTimeToNotify).String())
+		return reconcile.Result{RequeueAfter: requeueAfterTimeToNotify}, nil
+	}
+
+	// If the usersignup state hasn't been set to deactivating, then set it now
+	if !states.Deactivating(usersignup) {
+		states.SetDeactivating(usersignup, true)
+
+		logger.Info("setting usersignup state to deactivating")
+		if err := r.client.Update(context.TODO(), usersignup); err != nil {
+			logger.Error(err, "failed to update usersignup")
+			return reconcile.Result{}, err
+		}
+
+		// Requeue so that the deactivation due time can be calculated after the notification has been sent.
+		// The sequence of events from here are:
+		// 1. This controller has now set the UserSignup state to deactivating if it's not already set
+		// 2. Reconciliation is requeued
+		// 3. UserSignup controller picks up that the deactivating state has been set, and responds by:
+		//		a) creating a pre-deactivating notification for the user, and
+		//		b) setting the "deactivating notification created" status condition to true.
+		// 4. This  controller reconciles again, and if it doesn't find the notification created status condition as
+		//    expected, requeues again, otherwise:
+		// 5. This controller calculates the amount of time that has passed since the deactivating notification was sent,
+		//  based on the LastTransitionTime of the condition. If enough time has now passed, it sets the UserSignup to deactivated.
+
+		return reconcile.Result{Requeue: true, RequeueAfter: time.Duration(10) * time.Second}, nil
```

Comment:
stupid me :man_facepalming: I just realized that we don't watch the UserSignup CRD in the deactivation controller

Reply:
But this actually brings a question - since we modify the `UserSignup` resource several times, and we also rely on the status of the `UserSignup`, shouldn't we start watching that CRD as well?

Reply:
Thinking about it more and more, I believe that we should start watching the `UserSignup` CRD as well (as the secondary resources). This controller relies on the changes in the CRs, and it is actually an important part of the complex lifecycle of the `UserSignup` CRD. 
This would however mean that we would need to add a mapper between `UserSignup` & `MUR` to the watcher, but I really believe that it's worth it. In addition, such a mapper should be easy - just read the `UserSignup.Status.CompliantUsername`
WDYT @alexeykazakov @sbryzak

Reply:
This is exactly what is bothering me here (I mentioned it to you Matous when we were talking about this): not obvious relation between UserSignup and deactivation controllers.
But yes, I think it would make sense to start watching UserSignup CRD here too. However maybe it's better to merge this PR first and then create a separate one to switch to a watcher instead relying on re-queuing. But I would leave it to Shane to decide (one PR or two). We might face unexpected issues with the watcher so in that case we would already have a working solution and could just abandon that idea in worst case scenario.

Reply:
We are not waiting for a change in `UserSignup.Spec` though, we are waiting for a change to `UserSignup.Status` - will a status change still trigger another reconcile?

Reply:
It depends on the watcher you use. We can watch for new generations only (GenerationChangedPredicate) or for status only when the Status changed (OnlyUpdateWhenGenerationNotChanged: https://github.com/codeready-toolchain/toolchain-common/blob/a2981455c932c8d35d3106711f3ab60600f723d9/pkg/predicate/predicate.go#L17) or for everything if we want.

Reply:
Let's please get this one merged first, before looking at further refactorings.

Reply:
By the way, I agree with both of you that we should probably watch UserSignup also, instead of relying on the requeue mechanism.

Reply:
As I see it:
* it would simplify the flow
* make the code more clear
* follow the convention that we should watch secondary resources
* maybe also improve the stability of the flow/controller.

Is there any Jira issue to track this?

Reply:
I've created an issue for this: https://issues.redhat.com/browse/CRT-1070

Reply:
Thanks :+1: